docs: add contributing guide (#799)
janus-idp/backstage-plugins: CONTRIBUTING.md

As a contributor, here are the guidelines we would like you to follow:

- [Code of conduct](#code-of-conduct)
- [How can I contribute?](#how-can-i-contribute)
- [Using the issue tracker](#using-the-issue-tracker)
- [Submitting a Pull Request](#submitting-a-pull-request)
- [Coding rules](#coding-rules)
- [Working with the code](#working-with-the-code)

We also recommend that you read [How to Contribute to Open Source](https://opensource.guide/how-to-contribute).

## Code of conduct

Help us keep **Janus-IDP** open and inclusive. Please read and follow our [Code of Conduct](CODE_OF_CONDUCT.md).

## How can I contribute?

### Improve documentation

As a **Janus-IDP** user, you are the perfect candidate to help us improve our documentation: typo corrections, clarifications, more examples, etc. Take a look at the [documentation issues that need help](https://github.com/janus-idp/backstage-plugins/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Akind%2Fdocumentation).

Please follow the [Documentation guidelines](#documentation).

### Give feedback on issues

> Some issues are created without the information requested in the [Bug report guideline](#bug-report).

Reviewer: What about a sentence more like: "Issues that lack relevant information can be very difficult to track down and fix, please follow the [Bug reporting guidelines](#bug-report) to help make them easier to resolve."?

2. If you cloned a while ago, get the latest changes from upstream and update dependencies:

```bash
$ git checkout main
$ git pull upstream main
$ yarn install
```

3. Create a new topic branch (off the main project development branch) to contain your feature, change, or fix:

```bash
$ git checkout -b <topic-branch-name>
```

4. Make your code changes, following the [Coding rules](#coding-rules)
5. Push your topic branch up to your fork:

```bash
$ git push origin <topic-branch-name>
```

6. [Open a Pull Request](https://help.github.com/articles/creating-a-pull-request/#creating-the-pull-request) with a clear title and description.

**Tips**:

- For ambitious tasks, open a Pull Request as soon as possible with the `[WIP]` prefix in the title, in order to get feedback and help from the community.
- [Allow Janus-IDP maintainers to make changes to your Pull Request branch](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
  This way, we can rebase it and make some minor changes if necessary.
  All changes we make will be done in the new commit, and we'll ask for your approval before merging them.

> ## Coding rules

Reviewer 2: Are there any naming rules that should be added here? Like file, plugin or folder names?

Author: Good call, I'll add that. 👍

### Source code

To ensure consistency and quality throughout the source code, all code modifications must have:

- No [linting](#lint) errors
- A [test](#tests) for every possible case introduced by your code change
- **100%** test coverage
- [Valid commit message(s)](#commit-message-and-pull-request-guidelines)
- Documentation for new features
- Updated documentation for modified features

### Documentation

To ensure consistency and quality, all documentation modifications must:

- Refer to brand in [bold](https://help.github.com/articles/basic-writing-and-formatting-syntax/#styling-text) with proper capitalization, i.e. **GitHub**, **Red Hat**, **Backstage**, **npm**
- Prefer [tables](https://help.github.com/articles/organizing-information-with-tables) over [lists](https://help.github.com/articles/basic-writing-and-formatting-syntax/#lists) when listing key values, i.e. List of options with their description

> - Use [links](https://help.github.com/articles/basic-writing-and-formatting-syntax/#links) when you are referring to:
> - a **Janus-IDP** concept described somewhere else in the documentation, i.e. How to [contribute](CONTRIBUTING.md)
> - a third-party product/brand/service, i.e. Integrate with [GitHub](https://github.com)
> - an external concept or feature, i.e. Create a [GitHub release](https://help.github.com/articles/creating-releases)
> - a package or module, i.e. The [`@janus-idp/backstage-plugin-topology`](https://github.com/janus-idp/backstage-plugins/tree/main/plugins/topology) module
> - Use the [single backtick `code` quoting](https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code) for:
> - commands inside sentences, i.e. the `yarn new` command
> - programming language keywords, i.e. `function`, `async`, `String`
> - packages or modules, i.e. The [`@janus-idp/backstage-plugin-topology`](https://github.com/janus-idp/backstage-plugins/tree/main/plugins/topology) module
> - Use the [triple backtick `code` formatting](https://help.github.com/articles/creating-and-highlighting-code-blocks) for:
> - code examples
> - configuration examples
> - sequence of command lines

Reviewer 3: ```suggestion
- Use [links](https://help.github.com/articles/basic-writing-and-formatting-syntax/#links) when you are referring to:
  - a **Janus-IDP** concept described somewhere else in the documentation, i.e. How to [contribute](CONTRIBUTING.md)
  - a third-party product/brand/service, i.e. Integrate with [GitHub](https://github.com)
  - an external concept or feature, i.e. Create a [GitHub release](https://help.github.com/articles/creating-releases)
  - a package or module, i.e. The [`@janus-idp/backstage-plugin-topology`](https://github.com/janus-idp/backstage-plugins/tree/main/plugins/topology) module
- Use the [single backtick `code` quoting](https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code) for:
  - commands inside sentences, i.e. the `yarn new` command
  - programming language keywords, i.e. `function`, `async`, `String`
  - packages or modules, i.e. The [`@janus-idp/backstage-plugin-topology`](https://github.com/janus-idp/backstage-plugins/tree/main/plugins/topology) module
- Use the [triple backtick `code` formatting](https://help.github.com/articles/creating-and-highlighting-code-blocks) for:
  - code examples
  - configuration examples
  - sequence of command lines
```

| ------------ | ------------------------------------------------------------------------------------------------------ |
| **fix**      | A bug fix                                                                                              |
| **feat**     | A new feature                                                                                          |
| **chore**    | A change that doesn't affect package codebase but rather the tooling around                            |
| **docs**     | Documentation only changes                                                                             |
| **style**    | Changes that do not affect the meaning of the code (white space, formatting, missing semicolons, etc.) |
| **refactor** | A code change that neither fixes a bug nor adds a feature                                              |
| **perf**     | A code change that improves performance                                                                |
| **test**     | Adding missing tests or correcting existing tests                                                      |

> 

Reviewer 3: ```suggestion
| **revert**     | Reverts a previous commit                                                      |
```
Should we include `revert` in this table as well since it's being added in the `pr-semantic` workflow in this PR?

# Navigate to the newly cloned directory
$ cd backstage-plugins
# Assign the original repo to a remote called "upstream"
$ git remote add upstream git@github.com:tumido/backstage-plugins.git
# Install the dependencies
$ yarn install
```

The Backstage has multiple dependencies. To pass through their installation, make sure to follow the [isolated-vm](https://github.com/laverdet/isolated-vm#requirements) requirements.

### Lint

The [backstage-plugins](https://github.com/janus-idp/backstage-plugins) repository use [ESLint](https://eslint.org/) for linting and [Prettier](https://prettier.io) for formatting.

Formatting will be automatically verified and fixed by `lint-staged` on the commit.

Before pushing your code changes make sure there are no linting errors with `yarn lint` and `yarn prettier`.

**Tips**:

- Most linting errors can be automatically fixed with `yarn prettier:fix`.

### Tests

The [backstage-plugins](https://github.com/janus-idp/backstage-plugins) repository uses [Jest](https://jestjs.io/) for writing and running tests.

Before pushing your code changes make sure all **tests pass** and the **coverage is 100%**:

```bash
$ yarn test
```

## Releasing changes

> This repository defaults to a rapid release scheme where we rather release fast, on every PR merge than restrict ourselves by a strict release cadence and policy. This brings contributors the opportunity to see the direct impact of their contributions since they are released immediately after the merge. The release process itself is done via the [semantic-release](https://semantic-release.gitbook.io/semantic-release/) tool. In order for it to work properly, it requires contributors to follow a simple set of rules:

Reviewer: Missing word in the first sentence: "...where we **would** rather..."

Author: Great suggestion, thanks. This is better. 🙂

| ~~Patch~~ `vX.Y.🆙` Fix Release      | Type must be `fix`, `docs`, `refactor`, `style`, `revert` or `chore(deps)` (in this case the scope must match)      | `fix(topology): fix pipelinerun status icon color`          |
| ~~Minor~~ `vX.🆙.0` Feature Release  | Type must be `feat`, `perf`                                                                                         | `feat(kiali): add namespace selector`                       |
| ~~Major~~ `v🆙.0.0` Breaking Release | `<type>(<optional-scope>)!: <message>` Notice the `!` token. It must be present. The type and scope are irrelevant. | `feat(ocm)!: Relocate OCM config and support multiple hubs` |
| No release                           | Type `chore` (except when the scope is `deps`), or for any type when the scope is `no-release`                      |

## Creating new plugins

A particular case of contribution is implementing a new plugin. Before you start implementing a new one, please consult our issue tracker and make a [New plugin request](#new-plugin-request). Once you have a green light from the community, you can start implementing your plugin.

Start by running our plugin template wizard:

```bash
yarn new
```

> This will lead you through plugin category selection and set up all the required boilerplate. By default, this will only create a new standalone plugin within the `plugins` folder and will not touch the `packages` folder where we host an empty Backstage instance. You can bypass this behavior by running `janus-cli new` directly instead, but we discourage it. We aim to enable a standalone, self-contained plugin development workspace, however, this is not fully possible at this moment.

Reviewer: I think you can remove two commas from the last sentence "...standalone self-contained plugin development workspace, however this is not fully..."

2. **Plugin id**: This value is used in templating and determines the resulting package name and folder name. Please see the table above and look where the `<ID>` is used. As you can see, you don't have to specify the plugin role within the ID value, the role is added to the package name automatically.

Once you finish bootstrapping your plugin folder, please look up the upstream documentation. For example:

- [The structure of a plugin](https://backstage.io/docs/plugins/structure-of-a-plugin)
- For frontend plugins, see the general [plugin development guide](https://backstage.io/docs/plugins/plugin-development) and [composability system](https://backstage.io/docs/plugins/composability).
- For backend plugins, see the backend specifici [plugin development guide](https://backstage.io/docs/plugins/backend-plugin/).

### Develop a new plugin together with a local backstage instance

Backstage's support for standalone plugin development is minimal (especially for backend plugins), therefore we include a minimal test instance within this repository.

> 1. Install the plugin via `yarn workspace [app|backend] add @janus-idp/<PLUGIN_NAME>@*` (`@*` at the end ensures the package is always linked to the local package in the `plugins` folder)
> 2. Run `yarn start:backstage`

Reviewer 4: Could be useful to add some things that briefly tripped me up:

1. I missed the `backstage-plugin-` prefix when adding plugins the first time
2. I was doing a lot of annoying `git stash`es and "okay, what do I actually care about in the status output" before I tried to find a better way

````suggestion
1. Install the plugin via `yarn workspace [app|backend] add @janus-idp/<PLUGIN_NAME>@*`
   * Example: `yarn workspace app add "@janus-idp/backstage-plugin-nexus-repository-manager@*"`
   * `@*` at the end ensures the package is always linked to the local package in the `plugins` folder
1. Follow the plugin's install instructions
1. (Optional) Tell git to assume the modified files are unchanged, so that
   status/diff commands are representative of your current work:

   ```shell
   # Example, list any dev app files where you don't want changes tracked
   git update-index --assume-unchanged \
       app-config.yaml \
       packages/app/package.json \
       packages/app/src/components/catalog/EntityPage.tsx
   ```

1. Run `yarn start:backstage`
````

One other small hurdle I had getting started was that I needed to create example entities with appropriate annotations, and point Backstage at them. Those could maybe be checked into the repo?

Author: Hi @rmartine-ias, thanks for your suggestions. Sorry for a late response, I was on PTO. I'll add that 🙂 

Btw:
> I missed the backstage-plugin- prefix when adding plugins the first time

Yeah, that is a common thing... I hope this will be solved with https://github.com/janus-idp/backstage-plugins/pull/782 where we'll get custom templates/boilerplate for newly created plugins.

### New plugin request

Plugins are a great way to extend Backstage capabilities and are an integral part of this repository.
Please provide as much detail and context as possible and fill the information requested in the [plugin suggestion template](https://github.com/janus-idp/backstage-plugins/issues/new?labels=plugin%2Cstatus%2Ftriage&projects=janus-idp/2&template=plugin.yaml&title=%F0%9F%94%8C+Plugin%3A+).

## Submitting a Pull Request

Good pull requests, whether patches, improvements, or new features, are a fantastic help.
They should remain focused in scope and avoid containing unrelated commits.

**Please ask first** before embarking on any significant pull requests (e.g. implementing features, refactoring code), otherwise you risk spending a lot of time working on something that the project's maintainers might not want to merge into the project.

If you have never created a pull request before, welcome 🎉 😄.

> [Here is a great tutorial](https://opensource.guide/how-to-contribute/#opening-a-pull-request) on how to send one 🙂

Reviewer 3: ```suggestion
[Here is a great tutorial](https://opensource.guide/how-to-contribute/#opening-a-pull-request) on how to create one 🙂
```

4. Make your code changes, following the [Coding rules](#coding-rules)
5. Push your topic branch up to your fork:

```bash
$ git push origin <topic-branch-name>
```

6. [Open a Pull Request](https://help.github.com/articles/creating-a-pull-request/#creating-the-pull-request) with a clear title and description.

**Tips**:

- For ambitious tasks, open a Pull Request as soon as possible with the `[WIP]` prefix in the title, in order to get feedback and help from the community.
- [Allow Janus-IDP maintainers to make changes to your Pull Request branch](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
  This way, we can rebase it and make some minor changes if necessary.
  All changes we make will be done in the new commit, and we'll ask for your approval before merging them.

## Coding rules

### Source code

To ensure consistency and quality throughout the source code, all code modifications must have:

- No [linting](#lint) errors
- A [test](#tests) for every possible case introduced by your code change
- **100%** test coverage
- [Valid commit message(s)](#commit-message-and-pull-request-guidelines)
- Documentation for new features

> - Updated documentation for modified features

Reviewer 3: Do we want to ask for screenshots for features that modify the UI or introduce a new UI? 
This can potentially be added to the #documentation section, as part of the clear guide on how to use a plugin?

Author: I think we can improve this later on if we find ourselves in need of being more specific.

  - an external concept or feature, i.e. Create a [GitHub release](https://help.github.com/articles/creating-releases)
  - a package or module, i.e. The [`@janus-idp/backstage-plugin-topology`](https://github.com/janus-idp/backstage-plugins/tree/main/plugins/topology) module
- Use the [single backtick `code` quoting](https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code) for:
  - commands inside sentences, i.e. the `yarn new` command
  - programming language keywords, i.e. `function`, `async`, `String`
  - packages or modules, i.e. The [`@janus-idp/backstage-plugin-topology`](https://github.com/janus-idp/backstage-plugins/tree/main/plugins/topology) module
- Use the [triple backtick `code` formatting](https://help.github.com/articles/creating-and-highlighting-code-blocks) for:
  - code examples
  - configuration examples
  - sequence of command lines

When adding a new plugin, provide a clear guide into how to use it. This includes:

- Installation instructions
- Development setup, including how to set up all necessary third party integrations
- Comprehensively describe all available configuration options

> 

Reviewer 3: ```suggestion
- Screenshots or Videos showcasing UI elements of the plugin
```

It should contain any information about **Breaking Changes** and is also the place to reference GitHub issues that this commit **Closes**.

**Breaking Changes** should start with the word `BREAKING CHANGE:` with a space or two newlines.
The rest of the commit message is then used for this.

#### Examples

```commit
fix(pencil): stop graphite breaking when too much pressure applied
```

```commit
feat(pencil): add 'graphiteWidth' option

Fix #42
```

```commit
perf(pencil)!: remove graphiteWidth option

BREAKING CHANGE: The graphiteWidth option has been removed.

The default graphite width of 10mm is always used for performance reasons.
```

## Working with the code

### Set up the workspace

[Fork](https://guides.github.com/activities/forking/#fork) the project, [clone](https://guides.github.com/activities/forking/#clone) your fork, configure the remotes and install the dependencies:

```bash
# Clone your fork of the repo into the current directory
$ git clone git@github.com:$(gh config get user -h github.com)/backstage-plugins.git
# Navigate to the newly cloned directory
$ cd backstage-plugins
# Assign the original repo to a remote called "upstream"

> $ git remote add upstream git@github.com:tumido/backstage-plugins.git

Reviewer 3: ```suggestion
$ git remote add upstream git@github.com:janus-idp/backstage-plugins.git
```

# Install the dependencies
$ yarn install
```

The Backstage has multiple dependencies. To pass through their installation, make sure to follow the [isolated-vm](https://github.com/laverdet/isolated-vm#requirements) requirements.

### Lint

The [backstage-plugins](https://github.com/janus-idp/backstage-plugins) repository use [ESLint](https://eslint.org/) for linting and [Prettier](https://prettier.io) for formatting.

Formatting will be automatically verified and fixed by `lint-staged` on the commit.

> Before pushing your code changes make sure there are no linting errors with `yarn lint` and `yarn prettier`.

Reviewer 3: ```suggestion
Before pushing your code changes make sure there are no linting errors with `yarn lint` and `yarn prettier:check`.
```
